test/e2e/auth/node_authn.go

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/pkg/master/ports"
+	"k8s.io/kubernetes/test/e2e/framework"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = SIGDescribe("[Feature:NodeAuthenticator]", func() {
+
+	f := framework.NewDefaultFramework("node-authn")
+	var ns string
+	var nodeIPs []string
+	BeforeEach(func() {
+		ns = f.Namespace.Name
+
+		nodeList, err := f.ClientSet.CoreV1().Nodes().List(metav1.ListOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(len(nodeList.Items)).NotTo(Equal(0))
+
+		for _, nodeAddress := range nodeList.Items[0].Status.Addresses {
+			if len(nodeAddress.Address) > 0 {
+				if nodeAddress.Type == v1.NodeInternalIP || nodeAddress.Type == v1.NodeExternalIP {
+					nodeIPs = append(nodeIPs, nodeAddress.Address)
+				}
+			}
+		}
+
+		// make sure ServiceAccount admission controller is enabled, so secret generation on SA creation works
+		sa, err := f.ClientSet.CoreV1().ServiceAccounts(ns).Get("default", metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(len(sa.Secrets)).NotTo(Equal(0))
+	})
+
+	It("The kubelet's main port 10250 should fail with no credentials", func() {
+		for _, nodeIP := range nodeIPs {
+			// Anonymous authentication is disabled by default
+			pod := createNodeAuthTestPod(f, nodeIP, "")
+			Expect(true, strings.Contains(readPodLog(f, pod), "Unauthorized"))
+		}
+	})
+
+	It("The kubelet can delegate ServiceAccount tokens to the API server", func() {
+		By("create a new ServiceAccount for authentication")
+		falseValue := false
+		newSA := &v1.ServiceAccount{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: ns,
+				Name:      "node-auth-newSA",
+			},
+			AutomountServiceAccountToken: &falseValue,
+		}
+		sa, err := f.ClientSet.CoreV1().ServiceAccounts(ns).Create(newSA)
+		Expect(err).NotTo(HaveOccurred())
+		saSecret, err := f.ClientSet.CoreV1().Secrets(ns).Get(sa.Secrets[0].Name, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
```

Comment:
How about instead of doing this manually, just allow automount and read it from the volume:
```
curl ... -header "Authorization: Bearer `cat /var/run/secrets/kubernetes.io/serviceaccount/token`"
```

Reply:
Yeah, this is better. Good advice.

Reply:
I don't understand this.... what do you mean "will get rendered"?